modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java

```diff
@@ -0,0 +1,382 @@
+        }
+    }
+
+    /**
+     * @return {@link #testSnapshot} parameters.
+     */
+    private static List<TestData> testSnapshotData() {
+        return List.of(
+            new TestData(false, false),
+            new TestData(true, true),
+            new TestData(false, true),
+            new TestData(true, false)
+        );
+    }
+
+    /**
+     * Tests that a joining raft node successfully restores a snapshot.
+     *
+     * @param testData Test parameters.
+     * @throws Exception If failed.
+     */
+    @ParameterizedTest
+    @MethodSource("testSnapshotData")
+    public void testSnapshot(TestData testData) throws Exception {
+        // Set up a raft group service
+        RaftGroupService service = prepareRaftGroup();
+
+        doBeforeStop(service);
+
+        // Select any node that is not the leader of the group
+        JRaftServerImpl toStop = servers.stream()
+            .filter(server -> !server.localPeer(raftGroupId()).equals(service.leader()))
+            .findAny()
+            .orElseThrow();
+
+        // Get the path to that node's raft directory
+        Path serverDataPath = toStop.getServerDataPath(raftGroupId());
+
+        // Get the path to that node's RocksDB key-value storage
+        Path dbPath = getListenerPersistencePath(getListener(toStop, raftGroupId()));
+
+        int stopIdx = servers.indexOf(toStop);
+
+        // Remove that node from the list of servers
+        servers.remove(stopIdx);
+
+        // Shutdown that node
+        toStop.stop();
+
+        // Create a snapshot of the raft group
+        service.snapshot(service.leader()).get();
+
+        doAfterStop(service);
+
+        // Create another raft snapshot
+        service.snapshot(service.leader()).get();
+
+        if (testData.deleteFolder) {
+            // Delete a stopped node's raft directory and key-value storage directory
+            // to check if snapshot could be restored by the restarted node
+            IgniteUtils.deleteIfExists(dbPath);
+            IgniteUtils.deleteIfExists(serverDataPath);
+        }
+
+        if (testData.interactAfterSnapshot) {
+            // Interact with the raft group after the second snapshot to check if the restarted node would see these
+            // interactions after restoring a snapshot and raft logs
+            doAfterSnapshot(service);
+        }
+
+        // Restart the node
+        JRaftServerImpl restarted = startServer(stopIdx);
+
+        assertTrue(waitForTopology(cluster.get(0), servers.size(), 3_000));
+
+        BooleanSupplier closure = snapshotCheckClosure(restarted, testData.interactAfterSnapshot);
+
+        boolean success = waitForCondition(closure, 10_000);
+
+        assertTrue(success);
+    }
+
+    /**
+     * Interacts with raft group before the follower is stopped.
+     *
+     * @param service Raft group service.
+     * @throws Exception If failed.
+     */
+    public abstract void doBeforeStop(RaftGroupService service) throws Exception;
```

Comment:
I think that `beforeFollowerStop`, `afterFollowerStop`, `afterSnapshot` might be better names for these methods